File: java/src/main/java/com/sketches_oxide/quantiles/ReqSketch.java
package com.sketches_oxide.quantiles;

import com.sketches_oxide.NativeSketch;
import com.sketches_oxide.MergeableSketch;
import com.sketches_oxide.native.SketchOxideNative;

/**
 * ReqSketch - Relative Error Quantile Sketch with Extreme Accuracy at Tails.
 *
 * A quantile sketch designed for scenarios where accuracy at the extremes
 * (very low or very high quantiles) is critical. The sketch provides
 * zero error at rank 0 and rank 1, with gracefully degrading accuracy
 * toward the middle.
 *
 * <p><strong>Algorithm:</strong>
 * <ul>
 *   <li>Uses a hierarchy of compactors with varying sampling rates</li>
 *   <li>Higher compactor levels sample more aggressively</li>
 *   <li>Mode determines which extreme has higher accuracy</li>
 *   <li>Provides relative error guarantees that improve near extremes</li>
 * </ul>
 *
 * <p><strong>Accuracy Modes:</strong>
 * <ul>
 *   <li>HIGH_RANK_ACCURACY (0): Better accuracy for high quantiles (p99, p99.9)</li>
 *   <li>LOW_RANK_ACCURACY (1): Better accuracy for low quantiles (p1, p0.1)</li>
 * </ul>
 *
 * <p><strong>Use cases:</strong>
 * <ul>
 *   <li>SLA monitoring requiring extreme percentile accuracy</li>
 *   <li>Tail latency analysis (p99.9, p99.99)</li>
 *   <li>Anomaly detection at distribution extremes</li>
 *   <li>Financial risk metrics (VaR, CVaR)</li>
 * </ul>
 *
 * <p><strong>Compared to Other Sketches:</strong>
 * <ul>
 *   <li>vs KLL: Better accuracy at extremes, slightly more space</li>
 *   <li>vs DDSketch: Rank-based vs value-based error guarantees</li>
 *   <li>vs TDigest: More predictable error bounds</li>
 * </ul>
 *
 * <p><strong>Example:</strong>
 * <pre>
 * // Monitor tail latencies with high accuracy at p99+
 * try (ReqSketch sketch = new ReqSketch(12, ReqSketch.HIGH_RANK_ACCURACY)) {
 *     for (double latency : latencies) {
 *         sketch.update(latency);
 *     }
 *
 *     // High accuracy at these extreme quantiles
 *     System.out.println("p99: " + sketch.quantile(0.99));
 *     System.out.println("p99.9: " + sketch.quantile(0.999));
 *     System.out.println("p99.99: " + sketch.quantile(0.9999));
 * }
 * </pre>
 *
 * @see <a href="https://arxiv.org/abs/2004.01668">ReqSketch Paper</a>
 */
public final class ReqSketch extends NativeSketch implements MergeableSketch<ReqSketch> {

    /**
     * Mode for high accuracy at high ranks (p99, p99.9, etc.).
     * Use this when you care most about upper percentiles.
     */
    public static final int HIGH_RANK_ACCURACY = 0;

    /**
     * Mode for high accuracy at low ranks (p1, p0.1, etc.).
     * Use this when you care most about lower percentiles.
     */
    public static final int LOW_RANK_ACCURACY = 1;

    private final int k;
    private final int mode;

    /**
     * Create a new ReqSketch.
     *
     * @param k the accuracy parameter, controls space/accuracy tradeoff
     *          Higher values = more accurate but more space
     *          Typical values: 6-50, common default is 12
     * @param mode accuracy mode: {@link #HIGH_RANK_ACCURACY} or {@link #LOW_RANK_ACCURACY}
     * @throws IllegalArgumentException if k &lt;= 0 or mode is invalid
     * @throws RuntimeException if native memory allocation fails
     */
    public ReqSketch(int k, int mode) {
        if (k <= 0) {
            throw new IllegalArgumentException("k must be positive, got: " + k);
        }
        if (mode != HIGH_RANK_ACCURACY && mode != LOW_RANK_ACCURACY) {
            throw new IllegalArgumentException(
                    "mode must be HIGH_RANK_ACCURACY (0) or LOW_RANK_ACCURACY (1), got: " + mode);
        }

        this.k = k;
        this.mode = mode;
        this.nativePtr = SketchOxideNative.reqsketch_new(k, mode);

        if (this.nativePtr == 0) {
            throw new RuntimeException("Failed to allocate ReqSketch");
        }
    }

    /**
     * Create a ReqSketch with high rank accuracy mode.
     *
     * Convenience constructor for the common case of monitoring tail percentiles.
     *
     * @param k the accuracy parameter
     * @throws IllegalArgumentException if k &lt;= 0
     */
    public ReqSketch(int k) {
        this(k, HIGH_RANK_ACCURACY);
    }

    /**
     * Private constructor for deserialization.
     */
    private ReqSketch(long nativePtr, int k, int mode) {
        this.nativePtr = nativePtr;
        this.k = k;
        this.mode = mode;
    }

    /**
     * Add a value to the sketch.
     *
     * @param value the value to add
     * @throws IllegalStateException if the sketch has been closed
     */
    public void update(double value) {
        checkAlive();
        SketchOxideNative.reqsketch_update(nativePtr, value);
    }

    /**
     * Get the estimated value at a given quantile.
     *
     * Accuracy is highest near rank 0 and rank 1, degrading toward the middle.
     * The specific extreme with highest accuracy depends on the mode setting.
     *
     * @param q the quantile to query (0 &lt;= q &lt;= 1)
     * @return the estimated value at quantile q
     * @throws IllegalStateException if the sketch has been closed
     * @throws IllegalArgumentException if q not in [0, 1]
     */
    public double quantile(double q) {
        checkAlive();
        if (q < 0 || q > 1) {
            throw new IllegalArgumentException("quantile must be in [0, 1], got: " + q);
        }
        return SketchOxideNative.reqsketch_quantile(nativePtr, q);
    }

    /**
     * Merge another ReqSketch into this one.
     *
     * Both sketches must have the same k and mode parameters.
     *
     * @param other the sketch to merge into this one
     * @throws IllegalStateException if either sketch is closed
     * @throws IllegalArgumentException if parameters don't match
     * @throws NullPointerException if other is null
     */
    @Override
    public void merge(ReqSketch other) {
        checkAlive();
        if (other == null) {
            throw new NullPointerException("other cannot be null");
        }
        other.checkAlive();

        if (this.k != other.k || this.mode != other.mode) {
            throw new IllegalArgumentException(
                    "Cannot merge ReqSketches with different parameters: " +
                            String.format("(k=%d, mode=%d) vs (k=%d, mode=%d)",
                                    this.k, this.mode, other.k, other.mode));
        }

        SketchOxideNative.reqsketch_merge(this.nativePtr, other.nativePtr);
    }

    /**
     * Serialize the sketch to binary format.
     *
     * @return binary representation of the sketch
     * @throws IllegalStateException if the sketch has been closed
     */
    public byte[] serialize() {
        checkAlive();
        return SketchOxideNative.reqsketch_serialize(nativePtr);
    }

    /**
     * Deserialize a ReqSketch from binary format.
     *
     * @param data the serialized data from {@link #serialize()}
     * @return a new ReqSketch instance
     * @throws IllegalArgumentException if data is invalid or corrupt
     * @throws NullPointerException if data is null
     */
    public static ReqSketch deserialize(byte[] data) {
        if (data == null) {
            throw new NullPointerException("data cannot be null");
        }

        long ptr = SketchOxideNative.reqsketch_deserialize(data);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid or corrupt serialized ReqSketch data");
        }

        return new ReqSketch(ptr, 12, HIGH_RANK_ACCURACY);
    }

    /**
     * Get the k parameter.
     *
     * @return the k value used at creation
     */
    public int getK() {
        return k;
    }

    /**
     * Get the accuracy mode.
     *
     * @return {@link #HIGH_RANK_ACCURACY} or {@link #LOW_RANK_ACCURACY}
     */
    public int getMode() {
        return mode;
    }

    /**
     * Get a human-readable description of the mode.
     *
     * @return "HighRankAccuracy" or "LowRankAccuracy"
     */
    public String getModeDescription() {
        return mode == HIGH_RANK_ACCURACY ? "HighRankAccuracy" : "LowRankAccuracy";
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            freeNative();
            nativePtr = 0;
        }
    }

    @Override
    protected void freeNative() {
        if (nativePtr != 0) {
            SketchOxideNative.reqsketch_free(nativePtr);
        }
    }

    @Override
    public String toString() {
        if (nativePtr != 0) {
            return String.format("ReqSketch(k=%d, mode=%s)", k, getModeDescription());
        }
        return "ReqSketch(closed)";
    }
}

File: dotnet/SketchOxide/src/Quantiles/ReqSketch.cs
using System;
using SketchOxide.Native;

namespace SketchOxide.Quantiles;

/// <summary>
/// Relative Error Quantile (REQ) sketch for accurate quantile estimation.
///
/// Provides exact min (p0) and max (p100) values with high accuracy for intermediate
/// quantiles. The k parameter controls the tradeoff between accuracy and space usage.
/// </summary>
/// <remarks>
/// REQ sketch is particularly useful when exact extremes are required along with
/// accurate percentile estimates. Higher k values provide better accuracy but use more memory.
/// </remarks>
public sealed class ReqSketch : NativeSketch
{
    private readonly uint _k;

    /// <summary>
    /// Creates a new REQ sketch with the specified k parameter.
    /// </summary>
    /// <param name="k">Size parameter controlling accuracy. Must be greater than 0.
    /// Higher values give better accuracy but use more space. Typical values: 12-256.</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if k is 0.</exception>
    /// <exception cref="OutOfMemoryException">Thrown if native memory allocation fails.</exception>
    public ReqSketch(uint k)
    {
        if (k == 0)
            throw new ArgumentOutOfRangeException(nameof(k), k, "K must be greater than 0");

        _k = k;
        NativePtr = SketchOxideNative.reqsketch_new(k);

        if (NativePtr == 0)
            throw new OutOfMemoryException("Failed to allocate native ReqSketch");
    }

    /// <summary>
    /// Private constructor for deserialization.
    /// </summary>
    private ReqSketch(uint k, nuint ptr)
    {
        _k = k;
        NativePtr = ptr;
    }

    /// <summary>
    /// Gets the k parameter of this sketch.
    /// </summary>
    public uint K
    {
        get
        {
            CheckAlive();
            return _k;
        }
    }

    /// <summary>
    /// Updates the sketch with a new value.
    /// </summary>
    /// <param name="value">The value to add to the sketch.</param>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public void Update(double value)
    {
        CheckAlive();
        SketchOxideNative.reqsketch_update(NativePtr, value);
    }

    /// <summary>
    /// Returns the estimated value at the given rank (quantile).
    /// </summary>
    /// <param name="rank">The rank to query, in range [0, 1].
    /// 0 = minimum, 0.5 = median, 1 = maximum.</param>
    /// <returns>The estimated value at the given rank.</returns>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if rank is outside [0, 1].</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public double Query(double rank)
    {
        CheckAlive();
        if (rank < 0 || rank > 1)
            throw new ArgumentOutOfRangeException(nameof(rank), rank, "Rank must be in range [0, 1]");

        return SketchOxideNative.reqsketch_query(NativePtr, rank);
    }

    /// <summary>
    /// Gets the exact minimum value (p0).
    /// </summary>
    /// <returns>The exact minimum value.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public double Min() => Query(0);

    /// <summary>
    /// Gets the exact maximum value (p100).
    /// </summary>
    /// <returns>The exact maximum value.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public double Max() => Query(1);

    /// <summary>
    /// Gets the estimated median value (p50).
    /// </summary>
    /// <returns>The estimated median value.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public double Median() => Query(0.5);

    /// <summary>
    /// Serializes the sketch to a byte array.
    /// </summary>
    /// <returns>Serialized sketch bytes.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public byte[] Serialize()
    {
        CheckAlive();
        return SketchOxideNative.reqsketch_serialize(NativePtr, out _);
    }

    /// <summary>
    /// Deserializes a REQ sketch from a byte array.
    /// </summary>
    /// <param name="data">Serialized sketch bytes.</param>
    /// <param name="k">The k parameter used when creating the original sketch.</param>
    /// <returns>A new ReqSketch instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ArgumentException">Thrown if data is invalid.</exception>
    public static ReqSketch Deserialize(byte[] data, uint k)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        nuint ptr = SketchOxideNative.reqsketch_deserialize(data, (ulong)data.Length);
        if (ptr == 0)
            throw new ArgumentException("Failed to deserialize ReqSketch: invalid data");

        return new ReqSketch(k, ptr);
    }

    /// <summary>
    /// Returns a string representation of the sketch.
    /// </summary>
    public override string ToString()
    {
        if (IsDisposed)
            return "ReqSketch(disposed)";
        return $"ReqSketch(k={_k})";
    }

    /// <summary>
    /// Frees the native Rust instance.
    /// </summary>
    protected override void FreeNative()
    {
        if (NativePtr != 0)
        {
            SketchOxideNative.reqsketch_free(NativePtr);
            NativePtr = 0;
        }
    }
}
